### ISSUE TITLE
Look into replacing Backbone in the web UI

### ISSUE THREAD

**Issue by strugee** [pump-io/pump.io#1240]
Backbone is real nice, but I think we can come up with a simpler model. This issue is just for discussion - these are just my thoughts, _not_ plans for where the project is going to go.

Things we could do:

* Replace the view code with a React-style virtual DOM implementation - basically we'd use the existing Jade to render the entire web UI wholesale, diff it with the current DOM, then apply those changes, optionally all on a Web Worker. This would even let us keep state in a single global object (atom) instead of scattered across the codebase
* Replace Backbone's router with a library that did something similar, but focused _only_ on routing (or just write it ourselves, treating the URL as state - this might actually be pretty tiny but needs investigation)
* Replace Backbone's model code with Observables (maybe - this needs some more thinking as I'm not totally familiar with that part of the code)

See https://www.youtube.com/watch?v=okk0BGV9oY0 for where some of these ideas come from.

**Comment by bernhardreiter** [comment]
If you are looking into replacing backbone, you should consider https://vuejs.org/ (v2) before you do your own implementation only for pump.io. It has a virtual DOM implementation , can use pug (formerly known as jade) and is very modular. (vue-router and vuex will give you rounting and global state). 

Given the date on the video, vuejs.org v2 may be something that Joreteg might have liked.

**Comment by strugee** [comment]
@bernhardreiter thanks, I'll look into it. Although just so we're clear we'd be using libraries for virtual DOM diffing, etc. - I'm guessing the amount of pump-specific code would actually be pretty small

**Comment by bernhardreiter** [comment]
@strugee even it you are using components and libraries, it will be quite a bit of work to tie this together. So it depends if Vue's mixture is close to what is envisoned for pump.io.

**Comment by strugee** [comment]
@bernhardreiter I just don't think so. I could definitely be wrong, though. We'll see if we ever implement this.

**Comment by vxcamiloxv** [comment]
Other option is https://emberjs.com/, the problem with vue.js is mantained by on person and sponser by one company, but ember.js has a good active community development, is stable and also use virtual DOM.

**Comment by bernhardreiter** [comment]
@vxcamiloxv  vuejs.org is financially supported [by ~15 organisations and ~100 backers](https://vuejs.org/support-vuejs/). There is a [good active community beyond Evan](https://github.com/orgs/vuejs/people).  I cannot say for sure what is better for pump.io, as I don't have first hand experience with emberjs. There is [a comparison by the Vuejs folks](https://vuejs.org/v2/guide/comparison.html#Ember).

**Comment by vxcamiloxv** [comment]
Vue.js is good, I work with vue.js in some project, but I like the community development of ember.js, is a great example how work with a community project.

**Comment by strugee** [comment]
I looked into Ember a bit and it appears as though to get the most out of it you really have to use a template language designed for Ember, which Jade/Pug is not. So that's a non-starter unfortunately.

**Comment by vxcamiloxv** [comment]
Yes, the template is a problem, but you can think as non-monolithic app where the front end is total independent, and the frontend work totally over the API in major part.

**Comment by strugee** [comment]
Are you suggesting that we abandon server-side rendering? We want to go in the opposite direction from that (i.e. we want to be rendering _more_ things serverside). And we're just not going to rewrite all the templates again. Jade is easily good enough already.

**Comment by vxcamiloxv** [comment]
If you decide use vue.js you can use pug there, but the [microservice](http://microservices.io/patterns/microservices.html) arquitecture has many advantages like scalability and performance, if you want more render in server side vue.js will not make a big difference compared to backbone.js

**Comment by strugee** [comment]
pump.io is not going to become a microservice, ever. There's no reason to (i.e. it's architectural astronauting) and it adds a lot of complexity for node admins. We want to make the process of running a node easier, not harder.

The "render more server-side" thing is completely separate from this issue. The reason for this issue is not to improve server-side rendering, it's only about making the client-side JS simpler, because simpler is _always_ better.

**Comment by vxcamiloxv** [comment]
OK, vue.js has server-side rendering, support pug template engineer, but backbone.js i s more simple and minimalist, I have plant to re-implement pump.io frontend and vue.js has the better solution for pump

**Comment by vxcamiloxv** [comment]
Webpack has a great [feature](https://webpack.js.org/guides/code-splitting-async/#require-ensure-) for lazy-loading content (css,js etc), this feature allow load content on demand and improve the performance and you can make [prefetch](https://medium.com/@thejameskyle/react-loadable-2674c59de178) of content.

This [example](https://tailordev.fr/blog/2016/03/17/loading-dependencies-asynchronously-in-react-components/) is with react.js but work also with vue.js, or ember.js.

**Comment by strugee** [comment]
I've been looking into this a bit and I think we could hack together something pretty decent with https://github.com/tdumitrescu/virtual-jade, https://github.com/snabbdom/snabbdom, and maybe something like https://github.com/Raynos/main-loop? Might spend some time exploring this soon-ish.

**Comment by Serkan-devel** [comment]
**Why not vanilla.js?**

**Comment by strugee** [comment]
Are you being serious? Because then the codebase will become an unmanageable mess of spaghetti code. It'll have orders of magnitude more bugs.

**Comment by vxcamiloxv** [comment]
@strugee snabbdom looks good, but is unpopular compared with Vue.js:https://github.com/vuejs/vue and vue is based in snabbdom

**Comment by Serkan-devel** [comment]
@strugee I recommend you this https://www.youtube.com/watch?v=k7n2xnOiWI8
